shared/lib_battery_lifetime_nmc.h

```diff
@@ -0,0 +1,115 @@
+or promote products derived from this software without specific prior written permission.
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ARE DISCLAIMED.IN NO EVENT SHALL THE COPYRIGHT HOLDER, CONTRIBUTORS, UNITED STATES GOVERNMENT OR UNITED STATES
+DEPARTMENT OF ENERGY, NOR ANY OF THEIR EMPLOYEES, BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY,
+OR CONSEQUENTIAL DAMAGES(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
+OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#ifndef SAM_SIMULATION_CORE_LIB_BATTERY_LIFETIME_NMC_H
+#define SAM_SIMULATION_CORE_LIB_BATTERY_LIFETIME_NMC_H
+
+#include <memory>
+
+#include "lib_util.h"
+#include "lib_battery_lifetime_calendar_cycle.h"
+
+struct lifetime_nmc_state {
+    double q_relative_li;                // %, SEI degradation
+    double q_relative_neg;               // %, cycle degradation
+    double dq_relative_li_old;
+    double dq_relative_neg_old;
+    double DOD_max;
+    int n_cycles_prev_day;
+
+    // for complex cycling of battery, b1 = summation of b1_dt * dt_day over a day
+    // lifetime capacity updated after 24 hours elapse.
+
+    double b1_dt;
+    double b2_dt;
+    double b3_dt;
+    double c2_dt;
+
+    friend std::ostream& operator<<(std::ostream& os, const lifetime_nmc_state& p);
+};
+
+class lifetime_nmc_t : public lifetime_t {
+public:
+    lifetime_nmc_t(double dt_hr);
+
+    lifetime_nmc_t(std::shared_ptr<lifetime_params> params_pt);
+
+    lifetime_nmc_t(const lifetime_nmc_t& rhs);
+
+    lifetime_nmc_t &operator=(const lifetime_nmc_t& rhs);
+
+    lifetime_t *clone() override;
+
+    /// Execute the lifetime models given the current lifetime run index, capacity model, and temperature
+    void runLifetimeModels(size_t lifetimeIndex, bool charge_changed, double prev_DOD, double DOD, double T_battery) override;
+
+    double estimateCycleDamage() override;
+
+    void replaceBattery(double percent_to_replace) override;
+
+    /// Calculate negative electrode voltage from SOC
+    static double calculate_Uneg(double SOC);
+
+    /// Calculate open circuit voltage from SOC
+    static double calculate_Voc(double SOC);
+
+protected:
+
+    std::unique_ptr<lifetime_cycle_t> cycle_model;
+
+    /// Capacity degradation due to SEI
+    double runQli();
+
+    /// Capacity degradation due to cycles
+    double runQneg(double T_battery, double SOC);
+
+    void initialize();
+
+    /// NMC Kandler Smith parameters
+    double Ea_d0_1 = 4126.0;
```

Comment:
Which of these can safely be hardcoded and which need to be tied to SSC inputs? V_ref and Rug are especially suspicious given what a user can change in SSC.

Reply:
I'm not sure it's possible to make a connection between the voltage assumptions in this life model with the ones in SAM's voltage model. The V_ref parameter as chosen will determine how the other parameters have been fit. I'm not sure adjusting some of these basic assumptions will make sense / extrapolate well. Instead, treating it as a black-box where the inputs are battery operational parameters (SOC, cycle count, T, age) seems a safer bet.